MSTest: document testconfig.json (#43743)
dotnet/docs: docs/core/testing/unit-testing-mstest-configure.md

</RunSettings>
```

## testconfig.json

When running your tests with MSTest, you can use a `testconfig.json` file to configure the behavior of the test runner. The `testconfig.json` file is a JSON file that contains the configuration settings for the test runner. The file is used to configure the test runner and the test execution environment. For more information, refer to [Microsoft.Testing.Platform testconfig.json documentation](unit-testing-platform-config.md#testconfigjson).

> Starting with MSTest 3.7, you can also configure MSTest runs in the same configuration file. The sections below describe the settings that you can use in the `testconfig.json` file.

Reviewer: ```suggestion
Starting with MSTest 3.7, you can also configure MSTest runs in the same configuration file. The following sections describe the settings that you can use in the `testconfig.json` file.
```

### MSTest element

> MSTest settings are grouped by functionality that are described in the sections below.

Reviewer: ```suggestion
MSTest settings are grouped by functionality that are described in the sections that follow.
```

| Entry | Default | Description |
|-------|---------|-------------|
| orderTestsByNameInClass | false | If you want to run tests by test names both in Test Explorers and on the command line, set this value to **true**. |
| enableBaseClassTestMethodsFromOtherAssemblies | true | A value indicating whether to enable discovery of test methods from base classes in a different assembly from the inheriting test class. |
| classCleanupLifecycle | EndOfAssembly | If you want the class cleanup to occur at the end of the class, set it to **EndOfClass**. |

#### AssemblyResolution settings

| Entry | Default | Description |
|-------|---------|-------------|
| paths | None | You can specify paths to extra assemblies when finding and running unit tests. For example, use these paths for dependency assemblies that aren't in the same directory as the test assembly. You can specify a path in the shape `{ "path": "...", "includeSubDirectories": "true/false" }`. |

#### Deployment settings

| Entry | Default | Description |
|-------|---------|-------------|
| deleteDeploymentDirectoryAfterTestRunIsComplete | true | To retain the deployment directory after a test run, set this value to **false**. |

> | deployTestSourceDependencies | true |A value indicating whether the test source references are to be deployed. |

Reviewer: ```suggestion
| deployTestSourceDependencies | true | Indicates whether the test source references are to be deployed. |
```

| mapNotRunnableToFailed | true | A value indicating whether a not runnable result is mapped to failed test. |
| treatClassAndAssemblyCleanupWarningsAsErrors | false | To see your failures in class cleanups as errors, set this value to **true**. |
| treatDiscoveryWarningsAsErrors | false | To report test discovery warnings as errors, set this value to **true**. |

#### Timeout settings

| Entry | Default | Description |
|-------|---------|-------------|
| assemblyCleanup | 0 | Specify globally the timeout to apply on each instance of assembly cleanup method. |
| assemblyInitialize | 0 | Specify globally the timeout to apply on each instance of assembly initialize method. |
| classCleanup | 0 | Specify globally the timeout to apply on each instance of class cleanup method. |
| classInitialize | 0 | Specify globally the timeout to apply on each instance of class initialize method. |
| test | 0 | Specify globally the test timeout. |
| testCleanup | 0 | Specify globally the timeout to apply on each instance of test cleanup method. |
| testInitialize | 0 | Specify globally the timeout to apply on each instance of test initialize method. |

> | useCooperativeCancellation | false | When set to `true`, in case of timeout, MSTest will only trigger cancellation of the `CancellationToken` but will not stop observing the method. This is more performant but relies on user to correctly flow the token through all paths. |

Reviewer: ```suggestion
| useCooperativeCancellation | false | When set to `true`, in case of timeout, MSTest will only trigger cancellation of the `CancellationToken` but will not stop observing the method. This behavior is more performant but relies on the user to correctly flow the token through all paths. |
```

> [!NOTE]
> `[Timeout]` attribute specified on a method overrides the global timeout. For example, `[Timeout(1000)]` on a method marked with [AssemblyCleanup] will override the global `assemblyCleanup` timeout.

### Example *testconfig.json* file

> The following json shows the contents of a typical *.testconfig.json* file. Copy this code and edit it to suit your needs.

Reviewer: ```suggestion
The following JSON shows the contents of a typical *.testconfig.json* file. Copy this code and edit it to suit your needs.
```

</RunSettings>
```

## testconfig.json

> Starting with MSTest 3.7, when running your tests through MSTest runner (a.k.a Microsoft.Testing.Platform), you can use a `testconfig.json` file to configure the behavior of the test runner. The `testconfig.json` file is a JSON file that contains the configuration settings for the test runner. The file is used to configure the test runner and the test execution environment. For more information, refer to [Microsoft.Testing.Platform testconfig.json documentation](unit-testing-platform-config.md#testconfigjson).

Reviewer 2: Does this article describe just the parts that are useful to configure MSTest, and the rest should be found in the other doc that is linked (but the link does not work and points to MD file, rather then a learn article)?

If yes, then it is not obvious from the description. Especially this part confuses me: 

file to configure the behavior of the test runner.  .... The file is used to configure the test runner ... for more info see the link. 

From that I don't see what this article is about, when for test runner you redirected me to the link.

### MSTest element

MSTest settings are grouped by functionality that are described in the sections below.

> | Entry | Default | Description |

Reviewer 2: Is Execution header missing here?

Author: Nope it's some error in what was done and slipped through review. I am planning on fixing code and doing a follow up on doc but at least I was sure not to miss anything.

Reviewer 2: where is the error? The settings you list here are shown in the "execution" section in the example json file.

Reviewer 2: It is me who is confusing stuff here. I somehow thought these settings are the ones that are in the execution section, and they are not.

| deleteDeploymentDirectoryAfterTestRunIsComplete | true | To retain the deployment directory after a test run, set this value to **false**. |
| deployTestSourceDependencies | true |A value indicating whether the test source references are to be deployed. |
| enabled | true | If you set the value to **false**, deployment items that you specify in your test method aren't copied to the deployment directory. |

#### Output settings

| Entry | Default | Description |
|-------|---------|-------------|
| captureTrace | false | Capture text messages coming from the `Console.Write*`, `Trace.Write*`, and `Debug.Write*` APIs that will be associated to the current running test. |

#### Parallelism settings

| Entry | Default | Description |
|-------|---------|-------------|
| enabled | false | Enable test parallelization. |
| scope | class | The scope of parallelization. You can set it to `method`. The default, `class`, corresponds to running all tests of a given class sequentially but multiple classes in parallel. |
| workers | 0 | The number of threads/workers to be used for parallelization. The default value maps to the number of processors on the current machine. |

> #### Run settings

Reviewer 2: change this to execution because that is how the section is called?

Author: Good catch, I followed the "linter" recommendation on "execution" without paying attention to the impact.

```suggestion
#### Execution settings
```

</RunSettings>
```

## testconfig.json

> When running your tests with MSTest, you can use a `testconfig.json` file to configure the behavior of the test runner. The `testconfig.json` file is a JSON file that contains the configuration settings for the test runner. The file is used to configure the test runner and the test execution environment. For more information, refer to [Microsoft.Testing.Platform testconfig.json documentation](unit-testing-platform-config.md#testconfigjson).

Reviewer 3: My understanding is that the file name should be `<AssemblyName>.testconfig.json`. However, we have an MSBuild target in Microsoft.Testing.Platform.MSBuild that copies testconfig.json to `<AssemblyName>.testconfig.json`, right?

Do we want to document that? Are there valid supported scenarios where Microsoft.Testing.Platform.MSBuild won't be referenced and the user will have to precisely name the file as `<AssemblyName>.testconfig.json`?

> Starting with MSTest 3.7, you can also configure MSTest runs in the same configuration file. The sections below describe the settings that you can use in the `testconfig.json` file.

Reviewer 3: This is about https://github.com/microsoft/testfx/pull/3872/ right?

In unit-testing-platform-config.md, maybe the new `--config-file` should be documented?

| classCleanup | 0 | Specify globally the timeout to apply on each instance of class cleanup method. |
| classInitialize | 0 | Specify globally the timeout to apply on each instance of class initialize method. |
| test | 0 | Specify globally the test timeout. |
| testCleanup | 0 | Specify globally the timeout to apply on each instance of test cleanup method. |
| testInitialize | 0 | Specify globally the timeout to apply on each instance of test initialize method. |
| useCooperativeCancellation | false | When set to `true`, in case of timeout, MSTest will only trigger cancellation of the `CancellationToken` but will not stop observing the method. This is more performant but relies on user to correctly flow the token through all paths. |

> [!NOTE]
> `[Timeout]` attribute specified on a method overrides the global timeout. For example, `[Timeout(1000)]` on a method marked with [AssemblyCleanup] will override the global `assemblyCleanup` timeout.

### Example *testconfig.json* file

The following json shows the contents of a typical *.testconfig.json* file. Copy this code and edit it to suit your needs.

Each element of the file is optional because it has a default value.

```json

> {
>   "mstest": {

Author: ```suggestion
{
  "platformOptions": {
  }
  "mstest": {
```

    "execution": {
        "mapInconclusiveToFailed" : true,
        "disableAppDomain": true,
        "considerFixturesAsSpecialTests" : false,
    },
    "parallelism" : {
        "enabled": true,
        "scope": "method",
    },
    "output": {
        "captureTrace": false
    }
  }
}

> ```

Reviewer 3: For clarify, I'd add "platformOptions" to the example, and add comments in JSON to clarify that "platformOptions" are applicable to any testing framework that uses the new platform, while the "mstest" part is specific to MSTest.
